File: treewalker/src/token.c
#include "stdio.h"

#include "stdlib.h"
#include "declarations.h"
#include "token.h"


Token token_create(TokenType type, char* lexeme, void *literal) {
  Token token = {.type = type, .literal = literal, .lexeme = lexeme};
  return token;
}

void token_print(Token *token) {
  printf("Token\n");
  printf("\t%s", token_to_string(token->type));
  printf("\t%zu", token->line);
}


char *token_to_string(TokenType type) {
  switch (type) {
    case LEFT_PAREN:      return "LEFT_PAREN";
    case RIGHT_PAREN:     return "RIGHT_PAREN";
    case LEFT_BRACE:      return "LEFT_BRACE"; 
    case RIGHT_BRACE:     return "RIGHT_BRACE";
    case COMMA:           return "COMMA"; 
    case DOT:             return "DOT";          
    case MINUS:           return "MINUS";        
    case PLUS:            return "PLUS";        
    case SEMICOLON:       return "SEMICOLON";
    case SLASH:           return "SLASH";
    case STAR:            return "STAR";         
    case BANG:            return "BANG";         
    case BANG_EQUAL:      return "BANG_EQUAL";   
    case EQUAL:           return "EQUAL";        
    case EQUAL_EQUAL:     return "EQUAL_EQUAL"; 
    case GREATER:         return "GREATER";      
    case GREATER_EQUAL:   return "GREATER_EQUAL";
    case LESS:            return "LESS";         
    case LESS_EQUAL:      return "LESS_EQUAL";   
    case IDENTIFIER:      return "Identifier";
    case STRING:          return "String";
    case NUMBER:          return "Number"; 
    case AND:             return "and";
    case CLASS:           return "class";
    case ELSE:            return "else";
    case FALSE:           return "false";
    case FUN:             return "fun";
    case FOR:             return "for";
    case IF:              return "if";
    case NIL:             return "nil";
    case OR:              return "or";
    case PRINT:           return "print";
    case RETURN:          return "return";
    case SUPER:           return "super";
    case THIS:            return "this";
    case TRUE:            return "true";
    case VAR:             return "var";
    case WHILE:           return "while";
    case FILE_EOF:        return "eof";
  }
  

}

File: src/token.rs
use std::default;

use crate::object::LoxObject;
use crate::source_location::SourceLocation;

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenType,
    pub location: SourceLocation,
    pub lexeme: String,
    pub literal: Option<LoxObject>,
}

#[derive(PartialEq, Eq, Debug, Clone, Default)]
pub enum TokenType {
    // Single character tokens
    LeftBrace,
    RightBrace,
    LeftParen,
    RightParen,

    // One or two character tokens
    Comma,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    Dot,

    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,

    // Literals
    Identifier,
    String,
    Number,

    // Keywords
    Class,
    Function,
    Else,
    False,
    For,
    If,
    Nil,
    Or,
    And,
    Print,
    Return,
    Super,
    This,
    True,
    Var,
    While,

    #[default]
    EOF,
}

impl Token {
    pub fn new(
        kind: TokenType,
        location: SourceLocation,
        lexeme: String,
        literal: Option<LoxObject>,
    ) -> Self {
        Token {
            kind,
            location,
            lexeme,
            literal,
        }
    }

    pub fn keyword(keyword: &str) -> TokenType {
        match keyword {
            "and" => TokenType::And,
            "class" => TokenType::Class,
            "else" => TokenType::Else,
            "false" => TokenType::False,
            "for" => TokenType::For,
            "fun" => TokenType::Function,
            "if" => TokenType::If,
            "nil" => TokenType::Nil,
            "or" => TokenType::Or,
            "print" => TokenType::Print,
            "return" => TokenType::Return,
            "super" => TokenType::Super,
            "this" => TokenType::This,
            "true" => TokenType::True,
            "var" => TokenType::Var,
            "while" => TokenType::While,
            _ => TokenType::Identifier,
        }
    }
}
